### ISSUE TITLE
Continue training from backbone on multi-gpu

### ISSUE THREAD

**Issue by mateusz-wozny** [ultralytics/ultralytics#4098]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

**Detection Model**
I want to ask why in the case of using a single GPU in this code https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/model.py#L373-#L375 model is loaded before running `train` function for trainer but for multi GPU such a procedure is not performed (https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/dist.py#L36-#L37). I noticed that for multi GPU no loading of pre-trained model (if we don't want to resume training) causes errors when I want to continue training from the backbone in the multi GPU (for example, when the pre-trained model was trained for 3 classes and I want to train the new model for 30 classes). Also when I want to continue training from ckpt resume of training is performed i.e. if the pre-trained model was trained for 30 epochs new model start training from 31 epochs, not from 0. I noticed that this change helps
```python
def generate_ddp_file(trainer):
    """Generates a DDP file and returns its file name."""
    module, name = f'{trainer.__class__.__module__}.{trainer.__class__.__name__}'.rsplit('.', 1)

    content = f'''overrides = {vars(trainer.args)} \nif __name__ == "__main__":
    from {module} import {name}
    from ultralytics.utils import DEFAULT_CFG_DICT

    cfg = DEFAULT_CFG_DICT.copy()
    cfg.update(save_dir='')   # handle the extra key 'save_dir'
    trainer = {name}(cfg=cfg, overrides=overrides)
    if not overrides.get('resume'):  # manually set model only if not resuming
        self.trainer.setup_model()
    trainer.train()'''
    (USER_CONFIG_DIR / 'DDP').mkdir(exist_ok=True)
    with tempfile.NamedTemporaryFile(prefix='_temp_',
                                     suffix=f'{id(trainer)}.py',
                                     mode='w+',
                                     encoding='utf-8',
                                     dir=USER_CONFIG_DIR / 'DDP',
                                     delete=False) as file:
        file.write(content)
    return file.name
```

**Comment by glenn-jocher** [comment]
@mateusz-wozny hi there!

Thanks for your detailed question and insightful observations. The different loading procedures for single and multi GPU setups are handled by PyTorch. In the case of the single GPU training, PyTorch will automatically duplicate the model in GPU before initiating the training. 

In a multi-GPU setting, PyTorch creates child processes for each GPU and does not automatically duplicate the model across GPUs. Therefore, the model must be loaded in each child process separately before starting training. The differential loading process does not impact the training process itself or the final result, but it does ensure that the model is correctly loaded across multiple GPUs. 

Regarding your question about resuming training from checkpoints, YOLOv8 is designed to resume training from the last checkpoint. If you train a model for 30 epochs and then resume training, it will indeed start from epoch 31. This is because the state of the optimizer, along with the training progress, are saved in the checkpoint.

Again, thanks for pointing out the specific discrepancy in the pre-trained model load cases. It appears the provided snippet is aiming to manually preset the model when not resuming in a multi-GPU setting. We'll definitely consider this point for improving the library, and we appreciate your valuable contribution! We are always pushing to make YOLOv8 better, and users like you make that possible.

Best regards,
Glenn

**Comment by mateusz-wozny** [comment]
@glenn-jocher Thank you for your great explanations. Your observations are correct, I want to use a backbone (not including the classification head) pre-trained in multi-GPU settings for training a new model also on multi-GPU.

**Comment by glenn-jocher** [comment]
@mateusz-wozny, you're welcome and your approach is completely sensible. Leveraging a pre-trained backbone is a common practice aiming to take advantage of the learned representations, which tends to result in improved performances and/or less required training time.

When you're using a pre-trained backbone in a multi-GPU setting, the backbone should ideally already be duplicated across your GPUs before initiating training. PyTorch provides a mechanism for it, which creates separate child processes for each GPU and loads the model on each GPU separately. 

On the other hand, if you are trying to resume training from a checkpoint with a different number of classes, modifications need to be made to adapt the classification layer to the new number of classes. It's important to note that all previously trained model weights (including the classification layer) can still be useful as a starting point even when the number of classes is different. 

I hope this demystifies the process although please feel free to reach out if you have more questions or need further clarifications. We appreciate your contributions to improving YOLOv8!
